### ISSUE TITLE
CC0013 (user ternary) rule should be more careful with nullable types. (VB)

### ISSUE THREAD

**Issue by gigazet** [code-cracker/code-cracker#468]
for example, this:
    If value Is DBNull.Value Then
                    Return Nothing
    Else
                    Return CDate(value)
    End If
not equals to:
Return If(value Is DBNull.Value, Nothing, CDate(value))

first will return Nothing, but second will return DateTime.MinValue.

**Comment by giggio** [comment]
Bug confirmed. This is a simple test that makes it explicit:

````vb
Sub Main()
    Console.WriteLine(Bar())
    Console.WriteLine(Baz())
End Sub

Public Function Bar() As Date?
    If True Then
        Return Nothing
    Else
        Return CDate("")
    End If
End Function

Public Function Baz() As Date?
    Return If(True, Nothing, CDate(""))
End Function
````

The problem seems to be that the ternary `If` operator results in a `Date`, and not in a `Date?`, and Nothing will be coerced by VB to `DateTime.MinValue`.
To fix this, we need to check the return type, if it is a nullable, and one of the return expressions might return null, this will always fail, and the analyzer should not report a diagnostic.

**Comment by adraut** [comment]
I think I'm missing something here.  In what scenario is the result of the if block and the ternary different or an invalid syntax?  The following executes with the correct output for each call.

```
Sub Main()
        Dim IfBlockDateNothingResult = IfBlockDate(Nothing)
        Dim IfBlockDateSomethingResult = IfBlockDate(DateTime.Now)

        Dim TernaryDateNothingResult = TernaryDate(Nothing)
        Dim TernaryDateSomethingResult = TernaryDate(DateTime.Now)

        Dim IfBlockNullableDateNothingResult = IfBlockNullableDate(Nothing)
        Dim IIfBlockNullableDateSomethingResult = IfBlockNullableDate(DateTime.Now)

        Dim TernaryNullableDateNothingResult = TernaryNullableDate(Nothing)
        Dim TernaryNullableDateSomethingResult = TernaryNullableDate(DateTime.Now)
    End Sub

    Public Function IfBlockDate(value As Object) As Date
        If value Is Nothing Then
            Return Nothing
        Else
            Return CDate(value)
        End If
    End Function

    Public Function TernaryDate(value As Object) As Date
        Return If(value Is Nothing, Nothing, CDate(value))
    End Function

    Public Function IfBlockNullableDate(value As Object) As Date?
        If value Is Nothing Then
            Return Nothing
        Else
            Return CDate(value)
        End If
    End Function

    Public Function TernaryNullableDate(value As Object) As Date?
        Return If(value Is Nothing, Nothing, CDate(value))
    End Function
```

**Comment by giggio** [comment]
@adraut If you run the code I posted just above yours you will see different results.
The difference is that the ternary operator is casting to Date, and in VB, when you cast null to date it will produce DateTime.MinValue, which is then returned.

**Comment by adraut** [comment]
@giggio OK I see it now.

**Comment by giggio** [comment]
Should also fix assingment:

````vb
    Function foo() As Date?
        Dim d As DateTime?
        If True Then
            d = Nothing
        Else
            d = CDate(Nothing)
        End If
    End Function
````

**Comment by giggio** [comment]
Closed on #541.
